program/rust/src/accounts/score.rs

```diff
@@ -0,0 +1,181 @@
+            .unwrap_or(self.num_publishers);
+
+        if publisher_index == self.num_publishers {
+            if self.num_publishers == PC_MAX_PUBLISHERS as usize {
+                return Err(ProgramError::AccountDataTooSmall);
+            }
+
+            self.publishers[self.num_publishers] = publisher;
+            self.num_publishers += 1;
+        }
+
+        let symbol_index = self
+            .symbols
+            .iter()
+            .position(|&x| x == price_account)
+            .ok_or(ProgramError::InvalidArgument)?;
+
+        self.set_publisher_permission(publisher_index, symbol_index, true);
+        self.calculate_scores()
+    }
+
+    pub fn del_publisher(
+        &mut self,
+        publisher: Pubkey,
+        price_account: Pubkey,
+    ) -> Result<(), ProgramError> {
+        let publisher_index = self
+            .publishers
+            .iter()
+            .position(|&x| x == publisher)
+            .ok_or(ProgramError::InvalidArgument)?;
+
+        let symbol_index = self
+            .symbols
+            .iter()
+            .position(|&x| x == price_account)
+            .ok_or(ProgramError::InvalidArgument)?;
+
+        self.set_publisher_permission(publisher_index, symbol_index, false);
+        self.calculate_scores()
+    }
+
+    pub fn add_price(&mut self, symbol: Pubkey) -> Result<(), ProgramError> {
+        let symbol_index = self
+            .symbols
+            .iter()
+            .position(|&x| x == symbol)
+            .unwrap_or(self.num_symbols);
+
+        if symbol_index == self.num_symbols {
+            if self.num_symbols == PC_MAX_SYMBOLS as usize {
+                return Err(ProgramError::AccountDataTooSmall);
+            }
+
+            self.symbols[self.num_symbols] = symbol;
+            self.num_symbols += 1;
+        }
+
+        Ok(())
+    }
+
+    pub fn del_price(&mut self, symbol: Pubkey) -> Result<(), ProgramError> {
+        let symbol_index = self
+            .symbols
+            .iter()
+            .position(|&x| x == symbol)
+            .ok_or(ProgramError::InvalidArgument)?;
+
+        // update symbol list
+        self.symbols[symbol_index] = self.symbols[self.num_symbols - 1];
+        self.symbols[self.num_symbols - 1] = Pubkey::default();
+
+        // update publisher permissions
+        for i in 0..self.num_publishers {
+            let value = self.get_publisher_permission(i, self.num_symbols - 1);
+            self.set_publisher_permission(i, symbol_index, value)
+        }
+
+        self.num_symbols -= 1;
+        self.calculate_scores()
+    }
+
+    pub fn calculate_scores(&mut self) -> Result<(), ProgramError> {
+        let symbol_scores: Vec<u32> = self
+            .symbols
+            .iter()
+            .enumerate()
+            .map(|(j, _)| {
+                let score = self
+                    .publisher_permissions
+                    .iter()
+                    .fold(0, |score, permissions| score + permissions[j] as u32);
```

Comment:
I think this doesn't work well with your logic for storing the booleans in a u64s

Reply:
Fixed.